SI-9542 Fix regression in value classes (served two ways)

retronym opened the pull request:
```
% git log -3 --oneline --graph
*   b1c522a Merge branch 'ticket/9542' into ticket/9542-combo
|\
| * d09f024 SI-9542 Fix regression in value classes with enclosing refs
* | e7980e4 SI-9542 Unify different reprs. of module type refs
|/
```

retronym committed:
SI-9542 Fix regression in value classes with enclosing refs

In SI-9473 / e2653736, I changed `mkAttributedRef` to avoid prefixing
references to statically owned symbols with a `ThisType`.

Turns out there was one place that depended on the old behaviour.

```
object Outer {
  trait T
  class C(val value: Any) extends AnyVal {
	def foo(t: T) = expr
  }
}
```

Is translated to:

```
object Outer {
  trait T
  class C(val value: Any) extends AnyVal {
	def foo(t: T) = Outer.this.C.foo$extension(C.this, t)
  }
  object C {
    def foo$extension($this: C, t: T) = expr
  }
}
```

After the change, the forwarder was instead:

```
	def foo(t: T) = Outer.C.foo$extension(C.this, t)
```

Note: this change is not actually necessary after the following
commit that makes subtyping unify the different module class
reference, that alone is enough to make the enclosed test
pass.

retronym committed:
SI-9542 Unify different reprs. of module type refs

The new unit test shows failures in transitivity of subtyping
and type equivalence, which boil down the the inconsistent
handling of the semantically equivalent:

   ThisType(pre, ModuleClass)
   ModuleTypeRef(pre, ModuleClass)
   SingleType(pre, Module)

This commit:

   - adds a case to `normalizePlus` to unwrap a `ThisType` to
     a `ModuleTypeRef`
   - Use `normalizePlus` more widely during subtype comparison
   - refactor `fourthTry` (part of `isSubType`) to remove code
     that becomes obviated by the use of `normalizePlus`.

This fixes the regression in the extension methods phase which
was triggered by https://github.com/scala/scala/pull/4749.
We can also fix that regression by tweaking the extension methods
phase itself to emit the `ThisType` representation of the owner
of the value class, as before.

I plan to demonstrate the two approaches to fixing the regression
on separate branches, and the propose that the merged result of these
two is useds.

retronym commented:
Review by @adriaanm

adriaanm commented on src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala:
```diff
@@ -244,7 +244,7 @@ abstract class ExtensionMethods extends Transform with TypingTransformers {
           // These three lines are assembling Foo.bar$extension[T1, T2, ...]($this)
           // which leaves the actual argument application for extensionCall.
-          val sel        = Select(gen.mkAttributedRef(companion), extensionMeth)
+          val sel        = Select(gen.mkAttributedRef(companion.owner.thisType, companion), extensionMeth)
```
(based on hangout-review:) explain why this prefix is known to be valid (nesting is known by construction)

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -3967,14 +3967,15 @@ trait Types
-      case _                                       => tp.normalize
+      // Unify the representations of module classes
+      case st@SingleType(_, sym) if sym.isModule => st.underlying.normalize
+      case st@ThisType(sym) if sym.isModuleClass && sym.isStatic => normalizePlus(st.underlying)
```
document why `isStatic`?

retronym commented:
@adriaanm I've removed the `isStatic` condition (I can't remember why it was there, and will see if the test suite helps jog my memory). I've also added some more internal documentation to justify the use of `thistype` in `ExtensionMethods`.

adriaanm commented:
LGTM
